src/sacn/source.c

```diff
@@ -595,13 +1232,695 @@ etcpal_error_t sacn_source_reset_networking(sacn_source_t handle, uint16_t unive
+
+  EtcPalTimer interval_timer;
+  etcpal_timer_start(&interval_timer, SOURCE_THREAD_INTERVAL);
+
+  while (keep_running_thread || (num_thread_based_sources > 0))
+  {
+    num_thread_based_sources = process_internal(false);
+
+    etcpal_thread_sleep(etcpal_timer_remaining(&interval_timer));
+    etcpal_timer_reset(&interval_timer);
+
+    if (sacn_lock())
+    {
+      keep_running_thread = !shutting_down;
+      sacn_unlock();
+    }
+  }
+}
+
+// Takes lock
+int process_internal(bool process_manual)
+{
+  int num_sources_tracked = 0;
+
+  if (sacn_lock())
+  {
+    num_sources_tracked = process_sources(process_manual);
+    sacn_unlock();
+  }
+
+  return num_sources_tracked;
+}
+
+// Needs lock
+int process_sources(bool process_manual)
+{
+  int num_sources_tracked = 0;
+
+  // For each source
+  EtcPalRbIter source_iter;
+  etcpal_rbiter_init(&source_iter);
+  SacnSource* source = etcpal_rbiter_first(&source_iter, get_sacn_sources());
+  while (source)
+  {
+    bool source_incremented = false;
+
+    // If this is the kind of source we want to process (manual vs. thread-based)
+    if (source->process_manually == process_manual)
+    {
+      // If the Source API is shutting down, cause this source to terminate (if thread-based)
+      if (!process_manual && shutting_down)
+        set_source_terminating(source);
+
+      // Count the sources of the kind being processed by this function
+      ++num_sources_tracked;
+
+      // Universe processing
+      process_universe_discovery(source);
+      process_universes(source);
+
+      // Clean up this source if needed
+      if (source->terminating && (etcpal_rbtree_size(&source->universes) == 0))
+      {
+        remove_sacn_source(&source, &source_iter);
+        source_incremented = true;
+      }
+    }
+
+    if (!source_incremented)
+      source = etcpal_rbiter_next(&source_iter);
+  }
+
+  return num_sources_tracked;
+}
+
+// Needs lock
+void process_universe_discovery(SacnSource* source)
+{
+  // Send another universe discovery packet if it's time
+  if (!source->terminating && etcpal_timer_is_expired(&source->universe_discovery_timer))
+  {
+    send_universe_discovery(source);
+    etcpal_timer_reset(&source->universe_discovery_timer);
+  }
+}
+
+// Needs lock
+void process_universes(SacnSource* source)
+{
+  // For each universe of this source
+  EtcPalRbIter universe_iter;
+  etcpal_rbiter_init(&universe_iter);
+  SacnSourceUniverse* universe = etcpal_rbiter_first(&universe_iter, &source->universes);
+  while (universe)
+  {
+    bool universe_incremented = false;
+
+    // Unicast destination-specific processing
+    process_unicast_dests(source, universe);
+
+    // Either transmit start codes 0x00 & 0xDD, or terminate and clean up universe
+    if (universe->terminating)
+      process_universe_termination(source, universe, &universe_iter);
+    else
+      process_universe_null_pap_transmission(source, universe);
+
+    if (!universe_incremented)
```

Comment:
Good thinking handling the case of universes being removed while iterating, but it looks like the implementation isn't totally fleshed out here - you'll have to assign universe_incremented from the return values of the functions above, and probably change their paramters from SacnSourceUniverse* to SacnSourceUniverse**.

However, as mentioned elsewhere, etcpal_rbtree_remove() invalidates _all_ iterators, so you will have to find another solution for handling removal while iterating in the mem.c functions.

Reply:
rbtrees and iterators might be more trouble than they're worth here because of cases like this, you might consider changing them to arrays.

Reply:
Ah, darn, didn't know that. Yeah, only things I can think of are either moving to arrays, or changing it so that it makes a separate local rbtree of things it wants to delete, and then goes through and deletes them after the iteration. I'll keep thinking about which option is best.

Reply:
Also good catch on the implementation not being fleshed out. I must have missed that when refactoring.

Reply:
Arrays would seem to be the easier case, and perhaps we should start relying on them more than trees in general.  That invalidation of all iterators is really unfortuate.

Reply:
This has all been switched to arrays.

Reply:
Resolved in https://github.com/ETCLabs/sACN/pull/32/commits/9a88cc62e4c3fcb6011291027e0f8274c5237b9a.